File: cpp/4sum.cc
/*
 * Filename: 4sum.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 24 Jul 2015 02:06:00 PM EDT
 */

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
	public:
		vector<vector<int> > fourSum(vector<int>& nums, int target) {
			vector<vector<int> > res;
			sort(nums.begin(), nums.end());

			for (vector<int>::size_type i = 0; i < nums.size(); i++) {
				if (i && nums[i] == nums[i-1]) {
					continue;
				}

				for (vector<int>::size_type j = i+1; j < nums.size(); j++) {
					if (j-i-1 && nums[j] == nums[j-1]) {
						continue;
					}

					int k = j+1;
					int l = nums.size() - 1;

					while (k < l) {
						if (k-j-1 && nums[k] == nums[k-1]) {
							k++;
							continue;
						}

						if (l < nums.size() - 1 && nums[l] == nums[l+1]) {
							l--;
							continue;
						}

						int sum = nums[i] + nums[j] + nums[k] + nums[l];
						if (sum == target) {
							vector<int> ans;
							ans.push_back(nums[i]);
							ans.push_back(nums[j]);
							ans.push_back(nums[k]);
							ans.push_back(nums[l]);
							res.push_back(ans);
							k++;
							l--;
						} else if (sum > target) {
							l--;
						} else {
							k++;
						}
					}
				}
			}

			return res;
		}
};

File: java/4Sum.java
/*
 * Filename: 4Sum.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Thu 12 Mar 2015 03:36:24 PM EDT
 */

public class Solution {
	public List<List<Integer>> fourSum(int[] num, int target) {
		Arrays.sort(num);

		List<List<Integer>> ret = new ArrayList<List<Integer>>();

		int n = num.length;
		for (int i = 0; i < n-3; i++) {
			if (i > 0 && num[i] == num[i-1]) {
				continue;
			}

			for (int j = i+1; j < n-2; j++) {
				if (j != i+1 && num[j] == num[j-1]) {
					continue;
				}

				int a = num[i];
				int b = num[j];
				int k = j+1;
				int l = n-1;

				while (k < l) {
					if (k != j+1 && num[k] == num[k-1]) {
						k++;
						continue;
					}
					
					if (l != n-1 && num[l] == num[l+1]) {
						l--;
						continue;
					}

					int c = num[k];
					int d = num[l];
					int sum = a + b + c + d;

					if (sum == target) {
						List<Integer> tuple = new ArrayList<Integer>();
						tuple.add(num[i]);
						tuple.add(num[j]);
						tuple.add(num[k]);
						tuple.add(num[l]);
						ret.add(tuple);
						k++;
						l--;
					} else if (sum < target) {
						k++;
					} else {
						l--;
					}
				}
			}
		}

		return ret;
	}
}
